src/index.ts

```diff
@@ -874,7 +885,17 @@ class ThreadPool {
   }
 
   _maybeDrain () {
-    if (this.taskQueue.size === 0 && this.skipQueue.length === 0) {
+    /**
+     * Our goal is to make it possible for user space to use the pool
+     * in a way where always waiting === 0, since we want to avoid creating tasks that can't execute
+     * immediately in order to provide back pressure to the task source.
+     */
+    const maxCapacity = this.options.maxThreads * this.options.concurrentTasksPerWorker;
```

Comment:
I don't think the options object is immutable so it might be a good idea to recalculate and store this value.

Reply:
Maybe I'll be missing something, but this is not exposed through the `Piscina` interface, in fact, the `ThreadPool` is hidden under a private variable, so it should be safe enough and in theory not mutable (at least not at user-land).

Or do you refer to store the `maxCapacity` to avoid calculate it every time?

Reply:
I don't actually know where options comes from so maybe it's not a problem...

Reply:
Where the `pool` is being stored: https://github.com/piscinajs/piscina/blob/0af59c6b866a25cfda45fbd849a98e8ca3982606/src/index.ts#L925-L926

Where `options` are passed: https://github.com/piscinajs/piscina/blob/0af59c6b866a25cfda45fbd849a98e8ca3982606/src/index.ts#L984

It should be safe, I believe

Reply:
Actually... no... the `options` object comes from the `Piscina` constructor which is a user provided object and can be modified externally by the user.

Reply:
That's probably a general problem.

Reply:
Ah, got it. Yes, you are right, as preserve the object down the path. We'll need to either store it somewhere or make a deep copy of it to avoid external mutations.

Most likely it will deserve its own PR.
For now, I'll store the value within the `ThreadPool` to avoid side-effects.